Tools/cases_generator/optimizer_generator.py

```diff
@@ -145,6 +160,182 @@ def goto_label(self, goto: Token, label: Token, storage: Storage) -> None:
         self.out.emit(goto)
         self.out.emit(label)
 
+    def replace_opcode_if_evaluates_pure(
+        self,
+        tkn: Token,
+        tkn_iter: TokenIterator,
+        uop: CodeSection,
+        storage: Storage,
+        inst: Instruction | None,
+    ) -> bool:
+        skip_to(tkn_iter, "SEMI")
+
+        if self.original_uop.properties.escapes:
+            raise analysis_error(
+                f"REPLACE_OPCODE_IF_EVALUATES_PURE cannot be used with an escaping uop {self.original_uop.properties.escaping_calls}",
+                self.original_uop.body.open
+            )
+        emitter = OptimizerConstantEmitter(self.out, {}, self.original_uop, copy.deepcopy(self.stack))
+        emitter.emit("if (\n")
+        assert isinstance(uop, Uop)
+        input_identifiers = replace_opcode_if_evaluates_pure_identifiers(uop)
+        assert len(input_identifiers) > 0, "Pure operations must have at least 1 input"
```

Comment:
```suggestion
        if len(input_identifiers) > 0: raise analysis_error("Pure operations must have at least 1 input", tkn)
```

Never assert on inputs. Asserts should be for invariants.